### DISCUSSION TITLE
Disk "marked for removal" process is extremely slow

### DISCUSSION THREAD

**Discussion by stevew-Purdue** [moosefs/moosefs#461] [Q&A]
Hi,

I'm replacing some smaller disks with larger ones and am doing it one at a time per chunkserver by marking them in mfshdd.cfg with an '*'.  I don't have enough space on the other local disks in these chunkservers to use the '<' for local redistribution of the chunks.

The process is going incredibly slow.  I'm seeing about one chunk per hour being removed from the disks that are marked for removal.  It's for this reason that I often opt for just directly replacing the disks without first marking them for removal.  The process goes much faster although I have to endure a day or two of being undergoal.

Any ideas of what I can do to speed up the process?

My chunkservers are running MFS 3.0.115 and the master is running MFS 3.0.114 (I'll try to get it upgraded soon!).  I have four chunkservers and all files are set to a goal of two.

Thanks for any suggestions!
Steve

**Comment by chogata** [comment]
How are you figuring out how many chunks are copied from the "marked for removal" disk? Because the only way to do it is to check the increase in chunk numbers on other disks/chunkservers. MooseFS does NOT remove chunks from MFR drives, it just copies them to other chunkservers and when all are copied, says the disk is "ready" for removal. The only moment any deletion occurs is if a client writes to a chunk, that has a copy still on MFR and this copy is marked wrong version and eventually removed.

If you really do have a problem with speed, then I would need more info: how many chunkservers do you have in total, do you use labels, what are your replication settings in your mfsmaster.cfg etc.

Also - it's never a good idea to run chunkservers in higher version than the master. When you upgrade, you should always start with your CGI and metaloggers (if you have them), then master, then chunkservers, then clients (mounts).

**Reply by stevew-Purdue** [reply]
Okay, that makes sense.  Is there no way of determining the progress of the disk removal process?  Trying to keep track of the increase in chunk numbers on other chunkservers in a busy storage system doesn't seem to be very practical.

Thanks,
Steve

**Reply by oszafraniec** [reply]
In some way you will see overgoal chunks during "marked for removal" if you switch to "counts 'regular' hdd space and 'marked for removal' hdd space" on CGI. Overgoal chunks will be around the number of chunks on the drive you are removing. And keep in mind that after overgoal number will be equal chunks numer of the drive there will be some time that the drive is still "not ready". Don't know why, mayby @chogata can give us some teory to my practical observations ;)

**Reply by stevew-Purdue** [reply]
I switched to "counts only 'regular' hdd space" on the Info page of the CGI interface but I didn't see any option for making a change related to "'marked for removal' hdd space".  I don't see any overgoal chunks but I do see a number of undergoal chunks which is just slightly less than the number of chunks on the disk marked for removal as displayed on the Disks page.

**Reply by pkonopelko** [reply]
Hi Steve,

_All chunks state matrix_ view, as described in brackets, _counts 'regular' hdd space and 'marked for removal' hdd space_. It means that it takes marked for removal (MFR) disks into consideration while counting available copies of chunks. So if you have MFR disks, the number of overgoal chunks will be growing, since, as @chogata described, the chunks' copies existing on MFR disks do not get deleted, but are replicated to other available servers (btw usually from other available copies – the copies that exist on the MFR disks are used only if using other copies failed). The number of overgoal chunks in this view (if you don't have other undergoal chunks, e.g. related to a broken disk or existing due to other heavy replications, etc.) is the number of already replicated chunks – extra copies created due to marking some disk(s) for removal. 

What may actually be helpful in your case is **_Regular chunks state matrix_** which _counts only 'regular' hdd space_. This view shows a state of the system as if all the MFR disks were _already disconnected_. This is why you see undergoal chunks in this view after marking some disk(s) for removal. If you do not have other (meaning: not related to MFR disks, especially undergoal) replications ongoing, you can assume that number of undergoal chunks in the "Regular" view is actually number of chunks that still need to be replicated before MFR disks can be safely disconnected (i.e. that need to be replicated to fulfil desired Goal with number of valid copies which would exist after removing the MFR disks). The number of undergoal chunks should be decreasing constantly as your chunks from MFR disks are being replicated.

To sum up: If the number of undergoal chunks in "Regular" Chunks State Matrix view is 0, you can safely disconnect MFR disks from Chunkserver (comment them out in `mfshdd.cfg` and reload the Chunkserver process) and you shouldn't have any undergoal chunks neither in "All" nor in "Regular" views after disconnecting.

**Reply by stevew-Purdue** [reply]
Thanks!  Your explanation, plus the previous comments from others, is a tremendous help.

Steve

**Reply by pkonopelko** [reply]
Thanks @stevew-Purdue :) By the way, regarding speed – you can increase the speed of chunk replications simply by editing the replication limits (`CHUNKS_WRITE_REP_LIMIT` and `CHUNKS_READ_REP_LIMIT` variables in `mfsmaster.cfg`), and, of course, reloading the `mfsmaster` process afterwards. These limits apply to replications related to MFR disks too. In your case increasing the 2nd value in each limit should be enough.

**Reply by stevew-Purdue** [reply]
Okay, I'll try increasing the replication limits a little.  Thanks!
